cpp/ql/src/Likely Bugs/Format/NonConstantFormat.ql

```diff
@@ -16,135 +24,114 @@
-  isIndirect = false and
-  exists(Type unspecified |
-    unspecified = t.getUnspecifiedType() and
-    not unspecified instanceof UnknownType
-  |
-    unspecified instanceof BuiltInType or
-    unspecified instanceof IntegralOrEnumType
+predicate isNonConst(DataFlow::Node node) {
+  node instanceof FlowSource
+  or
+  // Parameters of uncalled functions that aren't const
+  exists(UncalledFunction f, Parameter p |
+    f.getAParameter() = p and
+    p = node.asParameter()
   )
-}
-
-predicate isNonConst(DataFlow::Node node, boolean isIndirect) {
-  exists(Expr e |
-    e = node.asExpr() and isIndirect = false
-    or
-    e = node.asIndirectExpr() and isIndirect = true
-  |
-    exists(FunctionCall fc | fc = e |
-      not (
-        whitelistFunction(fc.getTarget(), _) or
-        fc.getTarget().hasDefinition()
-      )
-    )
-    or
-    exists(Parameter p | p = e.(VariableAccess).getTarget() |
-      p.getFunction().getName() = "main" and p.getType() instanceof PointerType
-    )
-    or
-    e instanceof CrementOperation
-    or
-    e instanceof AddressOfExpr
-    or
-    e instanceof ReferenceToExpr
-    or
-    e instanceof AssignPointerAddExpr
-    or
-    e instanceof AssignPointerSubExpr
-    or
-    e instanceof PointerArithmeticOperation
-    or
-    e instanceof FieldAccess
-    or
-    e instanceof PointerDereferenceExpr
-    or
-    e instanceof AddressOfExpr
-    or
-    e instanceof ExprCall
-    or
-    e instanceof NewArrayExpr
+  or
+  // Consider as an input any out arg of a function or a function's return where the function is not:
+  // 1. a function with a known dataflow or taintflow from input to output and the `node` is the output
+  // 2. a function where there is a known definition
+  // i.e., functions that with unknown bodies and are not known to define the output through its input
+  //       are considered as possible non-const sources
+  // The function's output must also not be const to be considered a non-const source
+  exists(Call c |
+    exists(Expr arg | c.getAnArgument() = arg | arg = node.asDefiningArgument())
     or
-    exists(Variable v | v = e.(VariableAccess).getTarget() |
-      v.getType().(ArrayType).getBaseType() instanceof CharType and
-      exists(AssignExpr ae |
-        ae.getLValue().(ArrayExpr).getArrayBase().(VariableAccess).getTarget() = v
-      )
-    )
+    c = node.asIndirectExpr()
+  ) and
+  not exists(FunctionInput input, FunctionOutput output, CallInstruction call |
+    // NOTE: we must include dataflow and taintflow. e.g., including only dataflow we will find sprintf
+    // variant function's output are now possible non-const sources
+    (
+      pragma[only_bind_out](call.getStaticCallTarget())
+          .(DataFlowFunction)
+          .hasDataFlow(input, output) or
+      pragma[only_bind_out](call.getStaticCallTarget()).(TaintFunction).hasTaintFlow(input, output)
+    ) and
+    node = callOutput(call, output)
+  ) and
```

Comment:
There's also a subtle thing to think about here: If `call.getStaticCallTarget()` does not have a result then the `not exist` will always be satisfied. Is that what we want?

I think the answer is 'no', and that we should move the call to `call.getStaticCallTarget()` out of the `not exists` so that `call.getStaticCallTarget()` must have a result.

With that suggestion, this will look like:
```ql
exists(Call c, Function func, CallInstruction call |
  exists(Expr arg | c.getAnArgument() = arg | arg = node.asDefiningArgument())
  or
  c = node.asIndirectExpr()
|
  func = call.getStaticCallTarget() and
  not exists(FunctionInput input, FunctionOutput output |
    // NOTE: we must include dataflow and taintflow. e.g., including only dataflow we will find sprintf
    // variant function's output are now possible non-const sources
    pragma[only_bind_out](func).(DataFlowFunction).hasDataFlow(input, output) or
    pragma[only_bind_out](func).(TaintFunction).hasTaintFlow(input, output)
  |
    node = callOutput(call, output)
  )
)
```

I also don't think we need to give a name to `FunctionInput input`. So we could trim this down even more to:
```ql
exists(Call c, Function func, CallInstruction call |
  exists(Expr arg | c.getAnArgument() = arg and arg = node.asDefiningArgument())
  or
  c = node.asIndirectExpr()
|
  func = call.getStaticCallTarget() and
  not exists(FunctionOutput output |
    // NOTE: we must include dataflow and taintflow. e.g., including only dataflow we will find sprintf
    // variant function's output are now possible non-const sources
    pragma[only_bind_out](func).(DataFlowFunction).hasDataFlow(_, output) or
    pragma[only_bind_out](func).(TaintFunction).hasTaintFlow(_, output)
  |
    node = callOutput(call, output)
  )
)
```

Finally, it's kinda weird that we have both a `Call c` and a `CallInstruction call` (I know we need to do this thing because you need to import the internal `ModelUtil` library). But we could clean it up slightly more by getting rid of `Call c` and only use `CallInstruction call` for our logic:
```ql
exists(Function func, CallInstruction call |
  exists(Expr arg | call.getPositionalArgumentOperand(_).getDef().getUnconvertedResultExpression() = arg and
  arg = node.asDefiningArgument()
  )
  or
  call.getUnconvertedResultExpression() = node.asIndirectExpr()
|
  func = call.getStaticCallTarget() and
  not exists(FunctionOutput output |
    // NOTE: we must include dataflow and taintflow. e.g., including only dataflow we will find sprintf
    // variant function's output are now possible non-const sources
    pragma[only_bind_out](func).(DataFlowFunction).hasDataFlow(_, output) or
    pragma[only_bind_out](func).(TaintFunction).hasTaintFlow(_, output)
  |
    node = callOutput(call, output)
  )
)
```
Note a subtle thing here. Previously you had `c.getAnArgument()` which doesn't get the qualifier, and so to preserve the current behavior I had to replace this with `call.getPositionalArgumentOperand(_).getDef()`. If you want to include the qualifier you can do `call.getAnArgument()`.

You may want to recheck whether the `only_bind_out`s are still needed (I suspect they are, but it's worth checking).

Reply:
Fixed